### ISSUE TITLE
`kind` field for Literal 

### ISSUE THREAD

**Issue by impinball** [estree/estree#61]
It is very useful for basic type checking, and IMO is a bit less hackish (and easier) than `typeof node.value`. Here's my idea: 

```js
interface Literal <: Node, Expression {
  type: "Literal";
  value: string | boolean | null | number | RegExp;
  // addition, the type
  kind: "string" |
    "boolean" |
    "null" |
    "undefined" |
    "number" |
    "regexp";
}
```

**Comment by sebmck** [comment]
I'm impartial to this since I wouldn't be able to rely on `kind` anyway in case the AST was mangled but I can very much see the benefits for other tooling. Also `undefined` shouldn't be in that list since it's an `Identifier`.

**Comment by impinball** [comment]
I filed a bug out of uncertainty (this would be a modification of the original spec).

**Comment by michaelficarra** [comment]
Add `"infinity"` to distinguish infinity literals from other numbers? This would allow SpiderMonkey AST to be serialised to JSON.

**Comment by gibson042** [comment]
"infinity" seems like an abuse of this proposal. That `Infinity` and `NaN` cannot appear directly in JSON is a problem analogous to representation of ES6 regular expression literals in ES<6 runtimes.

"kind" has value on its own, but is insufficient to cover portability, which I think requires either Literal subtypes or "raw".

**Comment by michaelficarra** [comment]
@gibson042 The regexp issue has already been solved by https://github.com/estree/estree/issues/27. `NaN` isn't a problem because there's no `NaN` literal.

**Comment by impinball** [comment]
I would also support a "raw" node as well.

**Comment by gibson042** [comment]
> `NaN` isn't a problem because there's no `NaN` literal.

Is the same not true of `Infinity`?

**Comment by sebmck** [comment]
@gibson042 

Nope, you can represent `Infinity` with a numeric literal:

```javascript
> 1e10000
Infinity
```

etc.

**Comment by gibson042** [comment]
Is that true in SpiderMonkey? esprima at least presents `value: null` in such cases, and represents both literal `Infinity` and `NaN` as Identifier.

Regardless, I maintain that `kind` is not the mechanism for addressing lossy representations.

**Comment by sebmck** [comment]
@gibson042 In acorn it does at least, sounds like an esprima bug.

**Comment by michaelficarra** [comment]
@gibson042: No, it doesn't. You're probably observing the JSON serialisation problem we're talking about. And `Infinity` and `NaN` *are* identifiers. Literal infinities are just very large numbers (around 2e308 and larger) that cannot be precisely represented by an IEEE double.

If you don't think `kind` is a good proposal, what backward-compatible alternative do you suggest?

**Comment by gibson042** [comment]
> You're probably observing the JSON serialisation problem we're talking about. 

D'oh, right you are. :flushed:

> If you don't think kind is a good proposal, what backward-compatible alternative do you suggest?

I _do_ think `kind` is a good proposal, just not `kind: "infinity"`.

**Comment by impinball** [comment]
I concur with `kind: "infinity"` being a bad idea, with backwards
compatibility being another concern.

**Comment by michaelficarra** [comment]
Backward compatibility with what? Who currently implements kind?

**Comment by impinball** [comment]
Identifier("Infinity") versus Literal(kind: "infinity"). Tools already
depend on the former. That clarify?

**Comment by sebmck** [comment]
@impinball We aren't proposing to parse the `Infinity` identifier as a `Literal`, that makes no sense. Infinite numeric literals would be given the `"infinity"` `kind` instead of `"number"`.

**Comment by impinball** [comment]
Oh. Thanks for clarifying, although I don't get exactly why such a
distinction is absolutely necessary, since infinity is a number, and
checking for an infinite literal specifically isn't that hard.

**Comment by sebmck** [comment]
You can't serialise an infinite numeric literal to JSON.

**Comment by RReverser** [comment]
What about `-Infinity`, then? Do we need to add `"-infinity"` type for it, too?

Also, I'm not sure JSON serialization should be something that drives design of AST itself since, when it's really needed, you can implement callback for `JSON.stringify` that handles these edge cases.

**Comment by impinball** [comment]
Adds to the list of reasons why I'm not sure `kind: "infinity"` is really
necessary. It's easy to check for, and in reality, `raw` is the best way to
serialize it to JSON, anyways.

**Comment by sebmck** [comment]
Yeah and checking the type of `value` is easy too.

**Comment by michaelficarra** [comment]
@RReverser there are no negative number literals in JavaScript. `-Infinity` must be generated using unary negation.

**Comment by RReverser** [comment]
@michaelficarra Depends on whether AST node was parsed or generated and inserted. In `{type: "Literal", value: num}` having negative `num` was never specified as an invalid state, and AST transformers might very well insert such nodes into the tree.

**Comment by impinball** [comment]
@RReverser Technically, a transformer or builder can still emit invalid ASTs. Nothing stops you from doing this, even though it's completely nonsensical.

```js
ConditionalExpression(
  IfStatement( // invalid
    Identifier("Foo"),
    null,
    BlockStatement([])
  ),
  BlockStatement([ // invalid
    NewExpression(
      Identifier("Foo"),
      []
    )
  ]),
  Literal(0)
)
```

```js
// Your invalid JS-ish result:
(if (Foo) {}) ? { new Foo(); } : 0
```

As for a negative numeric Literal, that would be an easy PR to make, although it would be breaking. Getting all the tools patched for that should be relatively easy, though. [The Literal type does not specifiy that restriction, by the way.](https://github.com/estree/estree/blob/master/spec.md#literal)

**Comment by impinball** [comment]
I do feel we are digressing from the topic, though. As for the current values (without `undefined`, which is parsed as an Identifier), is this okay? #63

**Comment by RReverser** [comment]
> The Literal type does not specifiy that restriction, by the way.

That's exactly what I was pointing to by "having negative num was never specified as an invalid state".

**Comment by sebmck** [comment]
Closed via #63.
